### ISSUE TITLE
Skip indexes and functions and dropped granules

### ISSUE THREAD

**Issue by den-crane** [ClickHouse/ClickHouse#52637]
-- granularity 100
```sql
ClickHouse local version 23.6.2.18 (official build).

create table test (K Int64,  A Int64, B Int64, C Int64, 
index x1 (greatest(A,B,C)) type minmax granularity 100) 
Engine=MergeTree order by K as select number,0,0,0 from numbers(1e7);

select count() from test where greatest(A,B,C) >0;

Index `x1` has dropped 15/1221 granules.


select count() from test where greatest(A,B,C)!=0;

Index `x1` has dropped 15/1221 granules.
```

why only 15 granules is dropped?

-- granularity 1
```sql
create table test (K Int64,  A Int64, B Int64, C Int64, 
index x1 (greatest(A,B,C)) type minmax granularity 1) 
Engine=MergeTree order by K as select number,0,0,0 from numbers(1e7);

select count() from test where greatest(A,B,C) >0;
Index `x1` has dropped 1221/1221 granules.
```

it seems it's a statistics/show issue.

```sql
create table test (K Int64,  A Int64, B Int64, C Int64, 
index x1 (A>0,B>0,C>0) type set(2) granularity 1000) 
Engine=MergeTree order by K as select number,0,0,0 from numbers(1e8);

select count() from test where A>0 and B>0 and C>0;

Index `x1` has dropped 14/12209 granules.
Selected 5/5 parts by partition key, 0 parts by primary key, 12209/12209 marks by primary key, 0 marks to read from 0 ranges
```

**Comment by den-crane** [comment]
set index is very slow for some reason

```
create table test (K Int64,  A Int64, B Int64, C Int64, 

index x1 (A>0,B>0,C>0) type set(2) granularity 1000

) Engine=MergeTree order by K as select number,0,0,0 from numbers(1e8);
Elapsed: 56.916 sec.


create table test (K Int64,  A Int64, B Int64, C Int64, 

index x1 (greatest(A,B,C)) type minmax granularity 1000

) Engine=MergeTree order by K as select number,0,0,0 from numbers(1e8);
Elapsed: 3.058 sec.
```
But why 20 times difference?

**Comment by wzb5212** [comment]
For the wrong skip granule number, I also encountered a similar problem, please check the pr I submitted, thanks.

**Comment by wzb5212** [comment]
I had a similar problem with the wrong number of skip granules.

**Comment by rschu1ze** [comment]
Fixed by #53616
